projects/meta_cl/networks/dendritic_networks.py

```diff
@@ -184,6 +182,116 @@ def conv_block(cls, in_channels, out_channels, kernel_size, stride, padding):
                 in_channels=in_channels, out_channels=out_channels,
                 kernel_size=kernel_size, stride=stride, padding=padding
             ),
+            nn.InstanceNorm2d(num_features=out_channels, affine=True),
+            nn.ReLU(),
+        ]
+
+    def apply_dendrites(self, y, dendrite_activations):
+        """
+        Apply dendrites as a gating mechanism.
+
+        TODO: Decide best method to apply dendrites.
+        """
+        # # Multiple by the sigmoid of the max along each segment.
+        # return y * torch.sigmoid(dendrite_activations.max(dim=2).values)
+
+        inds = dendrite_activations.abs().max(dim=2).indices
+        inds = inds.unsqueeze(dim=2)
+        dendrite_activations = torch.gather(dendrite_activations, dim=2, index=inds)
+        dendrite_activations = dendrite_activations.squeeze(dim=2)
+        dendrite_activations = torch.sigmoid(dendrite_activations)
+        return y * dendrite_activations
+
+    def forward(self, x):
+
+        mod = self.modulation(x)
+        pred = self.prediction(x)
+
+        dendrite_activations = self.segments(mod)
+        out = self.apply_dendrites(pred, dendrite_activations)
+
+        out = self.classifier(out)
+
+        return out
+
+
+class CloserToANMLDendriticNetwork(nn.Module):
+    """
+    An iteration of a dendritic network, based closely on `ANML`_. (The conv + max-pool
+    layers are identical). It's similar to ANMLDendriticNetwork, but does not include a
+    linear layer atop the modulation network. This makes it more similar to ANML's
+    original network. In fact, with `num_segments=1, dendrite_sparsity=0`, they should
+    be identical.
+
+    .. _ANML: nupic.research/projects/meta_cl/networks/anml_networks.py
+
+    Input shape expected is 3 x 28 x 28 (C x W x H) but an adaptive average pooling
+    layer helps accept other heights and width.
+
+    With default parameters and `num_classes-963`, it uses 5,132,832 weights-on
+    out of a total of 13,960,323 weights. In comparison, ANML uses 5,963,139 weights
+    and OML uses 5,172,675 weights. Thus, these is still room to add more.
+
+    The default parameters are chosen so that the number of on weights for this network
+    is close to that of ANMLDendriticNetwork.
+    """
+
+    def __init__(self, num_classes,
+                 num_segments=10,
```

Comment:
If it mimics ANML is the goal, shouldn't the default values be num_segments=1 and dendrite_sparsity=0, as mentioned in the docstring?

Reply:
That was a subgoal. Another goal was to create a network similar to ANML but one that had multiple segments. The network serves both of those goals.